runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -276,38 +341,53 @@ public RemoteBundle getBundle(
         throws Exception {
       // TODO: Consider having BundleProcessor#newBundle take in an OutputReceiverFactory rather
       // than constructing the receiver map here. Every bundle factory will need this.
-      ImmutableMap.Builder<String, RemoteOutputReceiver<?>> outputReceivers =
-          ImmutableMap.builder();
-      for (Map.Entry<String, Coder> remoteOutputCoder :
-          processBundleDescriptor.getRemoteOutputCoders().entrySet()) {
-        String outputTransform = remoteOutputCoder.getKey();
-        Coder coder = remoteOutputCoder.getValue();
-        String bundleOutputPCollection =
-            Iterables.getOnlyElement(
-                processBundleDescriptor
-                    .getProcessBundleDescriptor()
-                    .getTransformsOrThrow(outputTransform)
-                    .getInputsMap()
-                    .values());
-        FnDataReceiver outputReceiver = outputReceiverFactory.create(bundleOutputPCollection);
-        outputReceivers.put(outputTransform, RemoteOutputReceiver.of(coder, outputReceiver));
-      }
 
-      if (environmentExpirationMillis == 0) {
-        return processor.newBundle(outputReceivers.build(), stateRequestHandler, progressHandler);
+      if (environmentExpirationMillis == 0 && !loadBalanceBundles) {
+        return currentClient.processor.newBundle(
+            getOutputReceivers(currentClient.processBundleDescriptor, outputReceiverFactory)
+                .build(),
+            stateRequestHandler,
+            progressHandler);
       }
 
-      final WrappedSdkHarnessClient client =
-          environmentCaches.get(environmentIndex).getUnchecked(executableStage.getEnvironment());
-      client.ref();
+      final LoadingCache<Environment, WrappedSdkHarnessClient> currentCache;
+      if (loadBalanceBundles) {
```

Comment:
Just an idea: Instead of conditionally switching here, could we add an interface to handle the client selection? We could have two implementations, one regular, one for load balancing. Would make the code easier to read here (necessary data structures would go into the implementation) and also avoid runtime checks.

Reply:
I also had that idea, maybe as a follow-up.